Bottom: 3d12b4c775d154f3958c5ecbb0a8fdc3f4362330
Top:    68b2eea1d6a6b364c55c032e8b60e053d2ec602b
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-31 07:57:35 +0800

Refresh of vhost-don-t-use-extra-loop

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index ebc0c53..aa8a2ef 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -508,7 +508,7 @@ static void handle_tx(struct vhost_net *net)
 			head = __vhost_get_vq_desc(vq, vq->iov,
 						   ARRAY_SIZE(vq->iov),
 						   &out, &in, NULL, NULL,
-						vhost16_to_cpu(vq, indices[i]));
+				     vhost16_to_cpu(vq, vq->heads[i].id));
 			if (in) {
 				vq_err(vq, "Unexpected descriptor format for TX: "
 				       "out %d, int %d\n", out, in);
diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 0e6d43e..4bcd096 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2490,6 +2490,8 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	total = vq->avail_idx - vq->last_avail_idx;
 	ret = total = min(total, num);
 
+#if 0
+	/* FIXME: into heads directly */
 	while (total) {
 		/* FIXME: use u16 for vq->num */
 		copied = min((u16)(vq->num - last_avail_idx), total);
@@ -2507,9 +2509,17 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 
 	if (!heads)
 		goto out;
+#endif
 
-	for (i = 0; i < ret; i++)
-		heads[i].id = indices[i];
+	for (i = 0; i < ret; i++) {
+		ret2 = vhost_get_avail(vq, heads[i].id,
+				      &vq->avail->ring[last_avail_idx]);
+		if (unlikely(ret2)) {
+			vq_err(vq, "Failed to get descriptors\n");
+			return -EFAULT;
+		}
+		last_avail_idx = (last_avail_idx + 1) & (vq->num - 1);
+	}
 
 	total = ret;
 	last_used_idx = vq->last_used_idx & (vq->num - 1);
diff --git a/samples/bpf/xdp1_kern.c b/samples/bpf/xdp1_kern.c
index 2197421..a4b89d3 100644
--- a/samples/bpf/xdp1_kern.c
+++ b/samples/bpf/xdp1_kern.c
@@ -51,6 +51,8 @@ int xdp_prog1(struct xdp_md *ctx)
 	u64 nh_off;
 	u32 ipproto;
 
+	return rc;
+
 	nh_off = sizeof(*eth);
 	if (data + nh_off > data_end)
 		return rc;
